src/pdf/pdf-view.js

```diff
@@ -688,6 +696,41 @@ class PDFView {
 			highlightAll: this._findState.highlightAll,
 			findPrevious: true
 		});
+		this.a11yHandleSearchResultUpdate();
+	}
+
+
+	// After the search result is switched to, record which node the
+	// search result is in to place screen readers' virtual cursor on it
+	// + announce the result.
+	async a11yHandleSearchResultUpdate() {
+		await debounceUntilScrollFinishes(this._iframeWindow.document);
+		let searchResult = this._iframeWindow.document.querySelector(".highlight.selected.appended");
+		if (!searchResult || !this._findState.result) return;
+
+		this._options.setA11yVirtualCursorTarget(searchResult.parentNode);
+
+		let { index, total } = this._findState.result;
+		let { currentPageNumber } = this._iframeWindow.PDFViewerApplication.pdfViewer;
+		let pageLabel = this._getPageLabel(currentPageNumber - 1);
+
+		let span = searchResult.parentNode;
+		let previousSpanText = span.previousSibling?.textContent || "";
+		let nextSpanText = span.nextSibling?.textContent || "";
+		let snippet = previousSpanText + span.textContent + nextSpanText;
+
+		this._options.a11yAnnounceSearchMessage(index, total, pageLabel, snippet);
+	}
```

Comment:
I think `a11yAnnounceSearchMessage` might not be necessary at all, because whenever the current find result changes [here](https://github.com/zotero/reader/blob/6d411f174335e887ec58c46f2a7ee021720e15fe/src/common/reader.js#L819-L821), it can generate the message from it. Currently there is no snippet, but in future versions we can add it to find result.

Additionally, we had thoughts about adding a search result sidebar, similar to the one in Preview.app.

Reply:
I moved the search message announcement logic into the views because of the page label. While trying to actually use the search with a screen reader, I realized that the search result index announced now is not useful but what one needs is the page label and the snippet around the search result. Right now (as indicated by that feedback from the forums), it is still preferable for people to use screen reader's search functionality, which is not great since it wouldn't find content from pages that are not currently rendered.

So thinking that we need to at least include the page label, my initial attempt was to just modify the existing [setA11ySearchMessage](https://github.com/zotero/reader/blob/ffdddaa4841f23f98825257df865028910071d74/src/common/reader.js#L659C2-L659C28) by including the page label via `this._state.primaryViewStats.pageLabel` into the message and moving that function call into `onSetFindState` that you pointed out. So it looked like this: 
```
let onSetFindState = (params) => {
	  this._updateState({ [primary ? 'primaryViewFindState' : 'secondaryViewFindState']: params });
	  this.setA11ySearchResultMessage(primary);
  };
```
The issue was that `this._state.primaryViewStats.pageLabel` I was using in `this.setA11ySearchResultMessage` would return the label from the page that the last result was on, as opposed to the page of the new search result. Likely just a race condition when I was trying to fetch it before the state updates are done. A workaround would be some kind of timeout but that felt not reliable. So I thought we can just offload that onto the view where we can wait for the scroll to be done and grab `this._iframeWindow.PDFViewerApplication.pdfViewer.currentPageNumber` to tell ` a11yAnnounceSearchMessage` which page needs to be announced. 

Another benefit was that from the view, I could fetch the neighboring spans to make a snippet ourselves for the pdf pages.

So I suppose the main question is: is there a better way that I missed to get the page number of the new search result from `reader.js` to not have the views handle this?  Or to wait for the next state update, maybe, before getting `this._state.primaryViewStats.pageLabel`? If the search result sidebar is added, I think this might indeed become somewhat redundant and we'd just announce that to navigate search results, check the sidebar.

Reply:
We can make `primaryViewFindState.result` to include all the necessary current result data like snippet or page label. For now you can check if `primaryViewFindState.result.currentPageLabel` or `primaryViewFindState.result.currentSnippet` exist and use them, and otherwise don't include them. After merging [this](https://github.com/zotero/reader/pull/138#issuecomment-2332463322) I can make PDF view to actually include those properties into the result. Additionally, the current code that retrieves snippets will break, as I've completely rewritten how PDF search works in that branch to enable highlighting/underlining search result.

Reply:
Right, so I moved the `a11yAnnounceSearchMessage` back to Reader and I am calling it with a debounce in `onSetFindState`. It will add `currentPageLabel` and  `currentSnippet` from the updated `findState` if present. It used to be called from `findNext`/`findPrev` after delay, which I don't think was quite right at the very least because it doesn't fire on the very first search result. 

We can wait on for PDF view to return those properties. I assumed we'd also want epub and snapshot to include those properties in their search results as well so that everything is consistent. I added them for epub and snapshot, so those are already announced now. This part is obviously subject to review by @AbeJellinek.

This way, we could actually remove the entire `a11yHandleSearchResultUpdate` for snapshot and epub, since all we did there was set virtualCursorTarget and that can just be set in `findNext`/`findPrev` because it is known from the search result. For pdf view, I renamed it as `a11yWillPlaceVirtCursorOnSearchResult` to be more descriptive, and calling it from `setFindState` again just so it fires on the very first search result.

Quite a few pieces refactored here but I think it's way cleaner now than before!